src/Tools/Pagination/Paginator.php

```diff
@@ -51,6 +63,29 @@ public function __construct(
         $this->query = $query;
     }
 
+    /**
+     * Create an instance of Paginator with auto-detection of whether the provided
+     * query is suitable for simple (and fast) pagination queries, or whether a complex
+     * set of pagination queries has to be used.
+     */
+    public static function newWithAutoDetection(QueryBuilder $queryBuilder): self
+    {
+        /** @var array<string, Join[]> $joinsPerRootAlias */
+        $joinsPerRootAlias = $queryBuilder->getDQLPart('join');
+
+        // For now, only check whether there are any sql joins present in the query. Note,
+        // however, that it is doable to detect a presence of only *ToOne joins via the access to
+        // joined entity classes' metadata (see: QueryBuilder::getEntityManager()->getClassMetadata(className)).
```

Comment:
Performing this analysis on a QueryBuilder will be complex, as it has to account for joins done on previous joins and not on the root entity, replicating work already done by the query parser as part of the AST.